flytekit/types/numpy/ndarray.py

```diff
@@ -0,0 +1,74 @@
+import pathlib
+import typing
+from typing import Type
+
+import numpy as np
+
+from flytekit.core.context_manager import FlyteContext
+from flytekit.core.type_engine import TypeEngine, TypeTransformer, TypeTransformerFailedError
+from flytekit.models.core import types as _core_types
+from flytekit.models.literals import Blob, BlobMetadata, Literal, Scalar
+from flytekit.models.types import LiteralType
+
+
+class NumpyArrayTransformer(TypeTransformer[np.ndarray]):
+    """
+    TypeTransformer that supports np.ndarray as a native type.
+    """
+
+    NUMPY_ARRAY_FORMAT = "NumpyArray"
+
+    def __init__(self):
+        super().__init__(name="Numpy Array", t=np.ndarray)
+
+    def get_literal_type(self, t: Type[np.ndarray]) -> LiteralType:
+        return LiteralType(
+            blob=_core_types.BlobType(
+                format=self.NUMPY_ARRAY_FORMAT, dimensionality=_core_types.BlobType.BlobDimensionality.SINGLE
+            )
+        )
+
+    def to_literal(
+        self, ctx: FlyteContext, python_val: np.ndarray, python_type: Type[np.ndarray], expected: LiteralType
+    ) -> Literal:
+        meta = BlobMetadata(
+            type=_core_types.BlobType(
+                format=self.NUMPY_ARRAY_FORMAT, dimensionality=_core_types.BlobType.BlobDimensionality.SINGLE
+            )
+        )
+
+        local_path = ctx.file_access.get_random_local_path() + ".npy"
+        pathlib.Path(local_path).parent.mkdir(parents=True, exist_ok=True)
+
+        # save numpy array to a file
+        # allow_pickle=False prevents numpy from trying to save object arrays (dtype=object) using pickle
+        np.save(file=local_path, arr=python_val, allow_pickle=False)
```

Comment:
We can probably leverage [typing.Annotated](https://docs.python.org/3/library/typing.html#typing.Annotated) to offer an escape hatch in case the user wants to pickle object arrays (with all the risks/caveats). Not for this PR though.